### PR TITLE
Storage done

### PR DESCRIPTION
**Pull request by Akado2009**
- The storage implmenetation
- All the actions/events/state queries/replies
- All the testing is done with public/private keypairs and signature to demonstrate the work

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -1,41 +1,308 @@
...
+    /// * `verifier` - MUST differ from the claim's subject or issuer.
+    ///
+    /// # Arguments:
+    /// * `piece_id` - claim's id.
+    /// * `subject` - subject's public key.
+    /// * `hash` - the hash to check against.
+    fn check_claim(&mut self, subject: PublicKey, piece_id: PieceId, hash: [u8; 32]) {
+        // TODO!: Rewrite in rust
+        let mut status = false;
+        if self
```
Comment (shamilsan) [inline src/lib.rs:172]: Why to not write it in Rust immediately?

File: src/lib.rs [hunk 2]
```diff
@@ -1,41 +1,308 @@
...
+                status,
+            },
+            0,
+        )
+        .expect("IDENTITY: Error during replying with IdentityEvent::CheckedClaim");
     }
 }
+
+#[no_mangle]
+pub unsafe extern "C" fn init() {
```
Comment (shamilsan) [inline src/lib.rs:197]: Why `pub`? `unsafe` is not necessary too.
```suggestion
extern "C" fn init() {
```

File: tests/identity_tests.rs [hunk 3]
```diff
@@ -0,0 +1,465 @@
...
+#[test]
+fn validation_status_from_subject() {
+    let sys = System::new();
+    let id_program = init_identity(&sys, USER);
+    let subject_pair = Sr25519Pair::from_seed(&hex!(
+        "9D61B19DEFFD5A60BA844AF492EC2CC44449C5697B326919703BAC031CAE7F60"
+    ));
+
+    let mut hasher = Sha256::new();
+    hasher.update(b"Moscow");
```
Comment (shamilsan) [inline tests/identity_tests.rs:135]: Let's choose another word.

File: tests/utils.rs [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+                "Unreachable metastate reply for the IdentityStateQuery::ValidationStatus payload has occured"
+            )
+        }
+    }
+}
+
+fn maps_match<
+    T: Eq + gstd::Ord + gstd::Clone,
+    U: gstd::Clone + gstd::PartialEq<V>,
+    V: gstd::Clone,
```
Comment (shamilsan) [inline tests/utils.rs:263]: Aren't these traits in preamble?
```suggestion
    T: Eq + Ord + Clone,
    U: Clone + PartialEq<V>,
    V: Clone,
```

File: tests/utils.rs [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+}
+
+pub fn check_verifiers_state_utils(
+    id_program: &Program,
+    subject: PublicKey,
+    piece_id: PieceId,
+    verifiers: Vec<PublicKey>,
+) {
+    match id_program.meta_state(IdentityStateQuery::Verifiers(subject, piece_id)) {
+        gstd::Ok(IdentityStateReply::Verifiers(real_verifiers)) => {
```
Comment (shamilsan) [inline tests/utils.rs:207]:
```suggestion
        Ok(IdentityStateReply::Verifiers(real_verifiers)) => {
```

File: tests/utils.rs [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+) {
+    match id_program.meta_state(IdentityStateQuery::Claim(subject, piece_id)) {
+        gstd::Ok(IdentityStateReply::Claim(real_claim)) => {
+            if let Some(cl) = real_claim {
+                if claim != cl {
+                    panic!("IDENTITY: Claims differ");
+                }
+            } else {
+                panic!("IDENTITY: No such claim");
+            }
```
Comment (shamilsan) [inline tests/utils.rs:190]: We can use expect here.
```suggestion
            if claim != real_claim.expect("IDENTITY: No such claim") {
                panic!("IDENTITY: Claims differ");
            }
```

File: tests/utils.rs [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+
+// META-STATE
+pub fn check_user_claims_state_utils(
+    id_program: &Program,
+    subject: PublicKey,
+    claims: BTreeMap<PieceId, Claim>,
+) {
+    match id_program.meta_state(IdentityStateQuery::UserClaims(subject)) {
+        gstd::Ok(IdentityStateReply::UserClaims(real_claims)) => {
+            if !maps_match(&real_claims, &claims) {
```
Comment (shamilsan) [inline tests/utils.rs:164]: Why can't we just compare two `BTreeMap`s?

File: tests/identity_tests.rs [hunk 8]
```diff
@@ -0,0 +1,465 @@
...
+    let claim = Claim {
+        issuer: subject_pair.public().0,
+        issuer_signature: subject_pair.sign(claim_data.encode().as_slice()).0,
+        subject: subject_pair.public().0,
+        verifiers: BTreeMap::new(),
+        data: claim_data,
+    };
+
+    issue_claim_utils(&id_program, USER, claim.clone(), PIECE_ID, false);
+    let mut claims: BTreeMap<PieceId, Claim> = BTreeMap::new();
```
Comment (shamilsan) [inline tests/identity_tests.rs:440]: No need to set the type explicitly.
```suggestion
    let mut claims = BTreeMap::new();
```

File: src/lib.rs [hunk 9]
```diff
@@ -1,41 +1,308 @@
...
-    let payload = String::from_utf8(msg::load_bytes()).expect("Invalid handle message");
+const ZERO_KEY: [u8; 32] = [0; 32];
+const ZERO_SIGNATURE: [u8; 64] = [0; 64];
 
-    if payload == "Hello" {
-        msg::reply(b"World", 0).unwrap();
-    }
+#[derive(Debug, Default)]
+pub struct IdentityStorage {
+    pub owner_id: ActorId,
```
Comment (fluiderson) [inline src/lib.rs:11]: This field seems to be unused.
```suggestion
```

File: io/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+
+/// Initializes an identity storage.
+///
+/// # Requirements:
+/// * `owner_id` MUST be non-zero address
+///
+/// `owner_id` - is the owner of the contract.
+#[derive(Decode, Encode, TypeInfo)]
+pub struct InitIdentity {
+    pub owner_id: ActorId,
```
Comment (fluiderson) [inline io/src/lib.rs:173]: See the previous comment.

File: src/lib.rs [hunk 11]
```diff
@@ -1,41 +1,308 @@
...
+            IdentityStateReply::ValidationStatus(
+                identity
+                    .user_claims
+                    .get(&pkey)
+                    .expect("No such public key")
+                    .get(&piece_id)
+                    .expect("No such piece_id")
+                    .data
+                    .valid,
+            )
```
Comment (fluiderson) [inline src/lib.rs:270]: We try to not panic in the `meta_state` function and return default value if smth fails.
I suggest to return `false` on fail:
```suggestion
            let mut status = false;

            if let Some(user_claim) = identity.user_claims.get(&pkey) {
                if let Some(claim) = user_claim.get(&piece_id) {
                    status = claim.data.valid
                }
            }

            IdentityStateReply::ValidationStatus(status)
```
Or return `Option<bool>` with the `None` variant on fail:
```suggestion
            let mut status = None;

            if let Some(user_claim) = identity.user_claims.get(&pkey) {
                status = user_claim.get(&piece_id);
            }

            IdentityStateReply::ValidationStatus(status)
```

File: src/lib.rs [hunk 12]
```diff
@@ -1,41 +1,308 @@
...
+        IdentityStateQuery::Date(pkey, piece_id) => IdentityStateReply::Date(
+            identity
+                .user_claims
+                .get(&pkey)
+                .expect("No such public key")
+                .get(&piece_id)
+                .expect("No such piece_id")
+                .data
+                .issuance_date,
+        ),
```
Comment (fluiderson) [inline src/lib.rs:281]: See the previous comment.

File: src/lib.rs [hunk 13]
```diff
@@ -1,41 +1,308 @@
...
+                .user_claims
+                .get(&pkey)
+                .expect("No such public key")
+                .get(&piece_id)
+                .expect("No such piece_id")
+                .verifiers
+                .keys()
+                .cloned()
+                .collect(),
+        ),
```
Comment (fluiderson) [inline src/lib.rs:293]: See the previous comment.

File: src/lib.rs [hunk 14]
```diff
@@ -1,41 +1,308 @@
...
-use gstd::{debug, msg, prelude::*};
+use gstd::{msg, prelude::*, ActorId};
+use identity_io::*;
 
-#[no_mangle]
-unsafe extern "C" fn handle() {
-    debug!("handle()");
-    let payload = String::from_utf8(msg::load_bytes()).expect("Invalid handle message");
+const ZERO_KEY: [u8; 32] = [0; 32];
+const ZERO_SIGNATURE: [u8; 64] = [0; 64];
```
Comment (fluiderson) [inline src/lib.rs:7]: Can you explain why assertions with these constants are needed? I don't think they should be on the contract side.

File: io/src/lib.rs [hunk 15]
```diff
@@ -0,0 +1,174 @@
+#![no_std]
+
+use codec::{Decode, Encode};
+use gstd::{prelude::*, ActorId};
+use scale_info::TypeInfo;
```
Comment (fluiderson) [inline io/src/lib.rs:5]:
```suggestion
use gstd::{prelude::*, ActorId};
```
These traits have already been imported by `gstd::prelude::*`.

File: src/lib.rs [hunk 16]
```diff
@@ -1,41 +1,308 @@
...
+
+#[no_mangle]
+pub unsafe extern "C" fn init() {
+    let config: InitIdentity = msg::load().expect("Unable to decode InitIdentity");
+    if config.owner_id == ActorId::zero() {
+        panic!("IDENTITY: Owner MUST be non-zero address");
+    }
+    let id_storage = IdentityStorage {
+        owner_id: config.owner_id,
+        piece_counter: 1,
```
[the remark this answers is not in the document]
Comment (Akado2009) [reply src/lib.rs:204]: 1 seems to be more user-friendly to me

[the remark this answers is not in the document]
Comment (fluiderson) [reply src/lib.rs:204]: There is 0 pieces during an initialization, so, IMO, it's rather confusing.
Also, `piece_counter` is used for setting IDs for new pieces, and similar counters in other smart contracts start from 0.
For example:
- NFT:
https://github.com/gear-dapps/non-fungible-token/blob/22bd6cc992bb34186a5bda4af7fcc376936551b6/nft/src/lib.rs#L13
(`token_id` is set to 0 by `Default::default()`):
https://github.com/gear-dapps/non-fungible-token/blob/22bd6cc992bb34186a5bda4af7fcc376936551b6/nft/src/lib.rs#L34
- Escrow:
https://github.com/gear-dapps/escrow/blob/972e9a35425657d4c1a2566a63d98d363e144c7e/src/lib.rs#L50
(`id_nonce` is set to 0 by `Default::default()`):
https://github.com/gear-dapps/escrow/blob/972e9a35425657d4c1a2566a63d98d363e144c7e/src/lib.rs#L153

[the remark this answers is not in the document]
Comment (Akado2009) [reply src/lib.rs:163]: Accepted :)

File: src/lib.rs [hunk 17]
```diff
@@ -218,23 +163,23 @@ async fn main() {
...
-        IdentityAction::CheckClaim {
-            subject,
-            piece_id,
-            hash,
-        } => identity.check_claim(subject, piece_id, hash),
+        // IdentityAction::CheckClaim {
+        //     subject,
+        //     piece_id,
+        //     hash,
+        // } => identity.check_claim(subject, piece_id, hash),
```
Comment (fluiderson) [inline src/lib.rs:182]: Dead code.

File: src/lib.rs [hunk 18]
```diff
@@ -1,41 +1,255 @@
...
+        self.piece_counter += 1;
+        msg::reply(
+            IdentityEvent::ClaimIssued {
+                issuer,
+                subject,
+                piece_id: self.piece_counter - 1,
+            },
+            0,
+        )
+        .expect("IDENTITY: Error during replying with IdentityEvent::ClaimIssued");
```
Comment (fluiderson) [inline src/lib.rs:52]:
```suggestion
        msg::reply(
            IdentityEvent::ClaimIssued {
                issuer,
                subject,
                piece_id: self.piece_counter,
            },
            0,
        )
        .expect("IDENTITY: Error during replying with IdentityEvent::ClaimIssued");
        self.piece_counter += 1;
```

File: io/src/lib.rs [hunk 19]
```diff
@@ -0,0 +1,173 @@
...
+    Claim(Option<Claim>),
+    Verifiers(Vec<PublicKey>),
+    ValidationStatus(bool),
+    Date(u64),
+    CheckedClaim(PublicKey, PieceId, bool),
+}
+
+/// Initializes an identity storage.
+#[derive(Decode, Encode, TypeInfo)]
+pub struct InitIdentity {}
```
Comment (fluiderson) [inline io/src/lib.rs:173]:
```suggestion
pub struct InitIdentity;
```

**Review by fluiderson — APPROVED** [review]

File: src/lib.rs (hunk 2 above)
Comment (shamilsan) [reply src/lib.rs:147]: I don't see it is resolved.

File: src/lib.rs [hunk 20]
```diff
@@ -1,41 +1,251 @@
...
+        )
+        .expect("IDENTITY: Error during replying with IdentityEvent::VerifiedClaim");
+    }
+}
 
-        let res = program.send_bytes(42, "Let's start");
-        assert!(res.log().is_empty());
+#[no_mangle]
+pub unsafe extern "C" fn init() {
+    let _config: InitIdentity = msg::load().expect("Unable to decode InitIdentity");
```
Comment (shamilsan) [inline src/lib.rs:148]: Why to read the payload if it is not used?

**Review by shamilsan — APPROVED** [review]

File: src/lib.rs (hunk 2 above)
Comment (Akado2009) [reply src/lib.rs:147]: my bad, sorry.

**Pull request merged**